File: web/core/src/ziwei/chart.ts
/**
 * 紫微斗数排盘主流程
 *
 * `ziweiPaipan(input)` 是本模块唯一的入口，纯函数、不碰网络与 DOM，结果可直接 `JSON.stringify`
 */

import { EarthBranch } from "tyme4ts"

import { EARTH_BRANCHES, HEAVEN_STEMS } from "../birth/constants"
import { correctTime } from "../birth/time"
import type { Location, PaipanInput } from "../birth/types"
import { brightnessOf } from "./data/brightness"
import {
  ADJECTIVE_STARS,
  BODY_MASTER,
  branchIndex,
  LIFE_MASTER,
  MAJOR_STARS,
  MINOR_STARS,
  mod,
  PALACE_NAMES,
  stemIndex,
} from "./data/tables"
import {
  boShiOf,
  changShengOf,
  decadesOf,
  minorLimitAgesOf,
  mutationsOf,
} from "./fortune"
import { lunarBirthOf } from "./lunar"
import { bodyPalaceOf, bureauOf, lifePalaceOf, palaceStemOf } from "./palaces"
import { placeStars } from "./stars"
import type {
  ZiweiChart,
  ZiweiMutation,
  ZiweiOptions,
  ZiweiPalace,
  ZiweiStar,
} from "./types"

export const DEFAULT_ZIWEI_OPTIONS: ZiweiOptions = {
  useTrueSolarTime: false,
  useDaylightSaving: false,
  lateZiAsNextDay: false,
  maxAge: 100,
}

/** 合并选项，`undefined` 的键当作没传，未知键忽略 */
export function resolveZiweiOptions(
  partial?: Partial<ZiweiOptions>
): ZiweiOptions {
  const merged: ZiweiOptions = { ...DEFAULT_ZIWEI_OPTIONS }
  if (!partial) return merged
  for (const key of Object.keys(
    DEFAULT_ZIWEI_OPTIONS
  ) as (keyof ZiweiOptions)[]) {
    const value = partial[key]
    if (value !== undefined) {
      merged[key] = value as never
    }
  }
  return merged
}

/** 落在某宫的星，按清单顺序，带庙陷与生年四化 */
function starsAt(
  names: readonly string[],
  positions: Record<string, number>,
  branch: number,
  mutations: ZiweiMutation[]
): ZiweiStar[] {
  return names
    .filter((name) => positions[name] === branch)
    .map((name) => {
      const star: ZiweiStar = { name }
      const brightness = brightnessOf(name, branch)
      if (brightness) star.brightness = brightness
      const mutation = mutations.find((m) => m.star === name)?.mutation
      if (mutation) star.mutation = mutation
      return star
    })
}

export function ziweiPaipan(
  input: PaipanInput,
  partial?: Partial<ZiweiOptions>
): ZiweiChart {
  const options = resolveZiweiOptions(partial)
  const { effective, info } = correctTime(input, options)
  const lunar = lunarBirthOf(effective, options.lateZiAsNextDay)

  const yearStem = stemIndex(lunar.yearSixtyCycle[0]!)
  const yearBranch = branchIndex(lunar.yearSixtyCycle[1]!)
  const yang = yearStem % 2 === 0
  const male = input.gender === "male"
  // 阳男阴女顺行
  const forward = yang === male

  const lifePalace = lifePalaceOf(lunar.effectiveMonth, lunar.hourBranchIndex)
  const bodyPalace = bodyPalaceOf(lunar.effectiveMonth, lunar.hourBranchIndex)
  const bureau = bureauOf(palaceStemOf(yearStem, lifePalace), lifePalace)

  const positions = placeStars({
    yearStem,
    yearBranch,
    month: lunar.effectiveMonth,
    day: lunar.day,
    hour: lunar.hourBranchIndex,
    lifePalace,
    bodyPalace,
    forward,
    yang,
    bureau: bureau.number,
  })
  const mutations = mutationsOf(yearStem)
  const changSheng = changShengOf(bureau.number, forward)
  const boShi = boShiOf(positions.minor["禄存"]!, forward)
  const decades = decadesOf(lifePalace, forward, bureau.number, lunar.year)
  const minorLimits = minorLimitAgesOf(yearBranch, male, options.maxAge)

  const palaces: ZiweiPalace[] = EARTH_BRANCHES.map((branch, b) => {
    const index = mod(lifePalace - b, 12)
    const stem = HEAVEN_STEMS[palaceStemOf(yearStem, b)]!
    return {
      index,
      name: PALACE_NAMES[index]!,
      branch,
      stem,
      sixtyCycle: stem + branch,
      isBodyPalace: b === bodyPalace,
      majorStars: starsAt(MAJOR_STARS, positions.major, b, mutations),
      minorStars: starsAt(MINOR_STARS, positions.minor, b, mutations),
      adjectiveStars: starsAt(
        ADJECTIVE_STARS,
        positions.adjective,
        b,
        mutations
      ),
      changSheng: changSheng[b]!,
      boShi: boShi[b]!,
      decade: decades[b]!,
      minorLimitAges: minorLimits[b]!,
    }
  })

  const location: Location | undefined =
    input.location !== undefined ||
    input.longitude !== undefined ||
    input.latitude !== undefined
      ? {
          name: input.location,
          longitude: input.longitude,
          latitude: input.latitude,
        }
      : undefined

  return {
    name: input.name,
    gender: input.gender,
    input,
    options,
    location,
    time: {
      ...info,
      // 生肖按农历年支，与八字按立春分界不同
      zodiac: EarthBranch.fromName(EARTH_BRANCHES[yearBranch]!)
        .getZodiac()
        .getName(),
    },
    lunar: {
      year: lunar.year,
      yearSixtyCycle: lunar.yearSixtyCycle,
      month: lunar.month,
      leap: lunar.leap,
      day: lunar.day,
      effectiveMonth: lunar.effectiveMonth,
      hourBranch: EARTH_BRANCHES[lunar.hourBranchIndex]!,
      text: lunar.text,
    },
    yearStem: HEAVEN_STEMS[yearStem]!,
    yearBranch: EARTH_BRANCHES[yearBranch]!,
    yang,
    forward,
    lifePalace: EARTH_BRANCHES[lifePalace]!,
    bodyPalace: EARTH_BRANCHES[bodyPalace]!,
    bureau,
    lifeMaster: LIFE_MASTER[yearBranch]!,
    bodyMaster: BODY_MASTER[yearBranch]!,
    mutations,
    palaces,
  }
}

File: internal/ziwei/chart.go
package ziwei

import (
	"github.com/6tail/tyme4go/tyme"

	"github.com/liasica/kismet/internal/birth"
)

// starsAt 落在某宫的星，按清单顺序，带庙陷与生年四化；总是返回非 nil 切片
func starsAt(names []string, positions map[string]int, branch int, mutations []Mutation) []Star {
	out := make([]Star, 0)
	for _, name := range names {
		if positions[name] != branch {
			continue
		}
		star := Star{Name: name, Brightness: BrightnessOf(name, branch)}
		for _, m := range mutations {
			if m.Star == name {
				star.Mutation = m.Mutation
			}
		}
		out = append(out, star)
	}
	return out
}

// Paipan 排盘
//
// 纯计算，不碰网络与磁盘，结果可直接序列化。与 TS 包 @kismet/core 的 ziweiPaipan
// 逐字段一致，由 data/fixtures/ziwei-charts.json 约束
func Paipan(input birth.Input, options Options) (Chart, error) {
	corrected, err := birth.CorrectTime(input, options.UseTrueSolarTime, options.UseDaylightSaving)
	if err != nil {
		return Chart{}, err
	}
	lunar := lunarBirthOf(corrected.Effective, options.LateZiAsNextDay)

	cycle := []rune(lunar.yearSixtyCycle)
	yearStem := stemIndex(string(cycle[0]))
	yearBranch := branchIndex(string(cycle[1]))
	yang := yearStem%2 == 0
	male := input.Gender == birth.GenderMale
	// 阳男阴女顺行
	forward := yang == male

	lifePalace := lifePalaceOf(lunar.effectiveMonth, lunar.hourBranch)
	bodyPalace := bodyPalaceOf(lunar.effectiveMonth, lunar.hourBranch)
	bureau := bureauOf(palaceStemOf(yearStem, lifePalace), lifePalace)

	positions := placeStars(starContext{
		yearStem:   yearStem,
		yearBranch: yearBranch,
		month:      lunar.effectiveMonth,
		day:        lunar.day,
		hour:       lunar.hourBranch,
		lifePalace: lifePalace,
		bodyPalace: bodyPalace,
		forward:    forward,
		yang:       yang,
		bureau:     bureau.Number,
	})
	mutations := mutationsOf(yearStem)
	changShengNames := changShengOf(bureau.Number, forward)
	boShiNames := boShiOf(positions.minor["禄存"], forward)
	decades := decadesOf(lifePalace, forward, bureau.Number, lunar.year)
	minorLimits := minorLimitAgesOf(yearBranch, male, options.MaxAge)

	palaces := make([]Palace, 0, 12)
	for b, branch := range earthBranches {
		index := mod(lifePalace-b, 12)
		stem := heavenStems[palaceStemOf(yearStem, b)]
		palaces = append(palaces, Palace{
			Index:          index,
			Name:           palaceNames[index],
			Branch:         branch,
			Stem:           stem,
			SixtyCycle:     stem + branch,
			IsBodyPalace:   b == bodyPalace,
			MajorStars:     starsAt(majorStars, positions.major, b, mutations),
			MinorStars:     starsAt(minorStars, positions.minor, b, mutations),
			AdjectiveStars: starsAt(adjectiveStars, positions.adjective, b, mutations),
			ChangSheng:     changShengNames[b],
			BoShi:          boShiNames[b],
			Decade:         decades[b],
			MinorLimitAges: minorLimits[b],
		})
	}

	// 生肖按农历年支，与八字按立春分界不同
	yearBranchName, err := tyme.EarthBranch{}.FromName(earthBranches[yearBranch])
	if err != nil {
		return Chart{}, err
	}
	timeInfo := corrected.Info
	timeInfo.Zodiac = yearBranchName.GetZodiac().GetName()

	return Chart{
		Name:     input.Name,
		Gender:   input.Gender,
		Input:    input,
		Options:  options,
		Location: birth.LocationOf(input),
		Time:     timeInfo,
		Lunar: Lunar{
			Year:           lunar.year,
			YearSixtyCycle: lunar.yearSixtyCycle,
			Month:          lunar.month,
			Leap:           lunar.leap,
			Day:            lunar.day,
			EffectiveMonth: lunar.effectiveMonth,
			HourBranch:     earthBranches[lunar.hourBranch],
			Text:           lunar.text,
		},
		YearStem:   heavenStems[yearStem],
		YearBranch: earthBranches[yearBranch],
		Yang:       yang,
		Forward:    forward,
		LifePalace: earthBranches[lifePalace],
		BodyPalace: earthBranches[bodyPalace],
		Bureau:     bureau,
		LifeMaster: lifeMaster[yearBranch],
		BodyMaster: bodyMaster[yearBranch],
		Mutations:  mutations,
		Palaces:    palaces,
	}, nil
}
